addons/lunch/models/lunch_order.py

```diff
@@ -41,10 +41,128 @@ class LunchOrder(models.Model):
+                    check = bool(len(toppings) == 1 if quantity == '1' else toppings)
+                    if not check:
+                        raise ValidationError(errors[quantity] % label)
+
+    @api.model
+    def create(self, values):
+        lines = self._find_matching_lines({
+            **values,
+            'toppings': self._extract_toppings(values),
+        })
+        if lines:
+            # FIXME This will update multiple lines in the case there are multiple matching lines
+            # which should not happen trough the interface
+            lines.update_quantity(1)
+            return lines[:1]
+        return super().create(values)
+
+    def write(self, values):
+        merge_needed = 'note' in values or 'topping_ids_1' in values or 'topping_ids_2' in values or 'topping_ids_3' in values
+
+        # Only write on topping_ids_1 because they all share the same table and we don't want to remove all the records
+        # _extract_toppings will pop topping_ids_1, topping_ids_2 and topping_ids_3 from values
+        # This also forces us to invalidate the cache for topping_ids_2 and topping_ids_3 that could have changed
+        # through topping_ids_1 without the cache knowing about it
+        toppings = self._extract_toppings(values)
+        values['topping_ids_1'] = [(6, 0, toppings)]
+        self.invalidate_cache(['topping_ids_2', 'topping_ids_3'])
+
+        if merge_needed:
+            lines_to_deactivate = self.env['lunch.order']
+            for line in self:
+                matching_lines = self._find_matching_lines({
+                    'user_id': values.get('user_id', line.user_id.id),
+                    'product_id': values.get('product_id', line.product_id.id),
+                    'note': values.get('note', line.note or False),
+                    'toppings': toppings,
+                })
+                if matching_lines:
+                    lines_to_deactivate |= line
+                    # TODO Try to batch it, be careful there might be multiple matching lines
+                    # for the same order hence quantity should not always be line.quantity, but rather a sum
+                    matching_lines.update_quantity(line.quantity)
+            lines_to_deactivate.write({'active': False})
+            return super(LunchOrder, self - lines_to_deactivate).write(values)
+        return super().write(values)
+
+    @api.model
+    def _find_matching_lines(self, values):
+        domain = [
+            ('user_id', '=', values.get('user_id', self.default_get(['user_id'])['user_id'])),
+            ('product_id', '=', values.get('product_id', False)),
+            ('date', '=', fields.Date.today()),
+            ('note', '=', values.get('note', False)),
+        ]
+        toppings = values.get('toppings', [])
+        return self.search(domain).filtered(lambda line: (line.topping_ids_1 | line.topping_ids_2 | line.topping_ids_3).ids == toppings)
```

Comment:
I don't like this filtered call, but I cannot find any other elegant way to do this, is seems to be impossible using a simple domain